This Notebook contains attempts to use the geopy client to search for geographical coordinates for Toronto Neighbourhoods. This has been fruitless so far but further investigation may be worthwhile

# Clustering Toronto Neighbourhoods

## Load libraries

In [1]:
import pandas as pd

## Data Exctraction and Cleaning

Toronto neighbourhood information was extractred from [Wikipedia](https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M) directly using the pandas read_html funtion.

In [2]:
# Extract table from wikipedia using Pandas
table_df = pd.read_html('https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M')[0]

# Remove unassigned boroughs
table_df = table_df[table_df.Borough != 'Not assigned'].reset_index(drop=True)

table_df.head()

Postal Code           Borough                                Neighbourhood
0         M3A        North York                                    Parkwoods
1         M4A        North York                             Victoria Village
2         M5A  Downtown Toronto                    Regent Park, Harbourfront
3         M6A        North York             Lawrence Manor, Lawrence Heights
4         M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government

Above we can see, we have the correct columns and the first 5 rows of the data frame look correct. 

## Data checking and verifying

The DataFrame was checked to make sure no post codes were shared and that all had neighbourhood(s) assigned. Finally the shape of the Dataframe was shown. 

In [3]:
table_df['Postal Code'].value_counts().value_counts()

1    103
Name: Postal Code, dtype: int64

Here we can see we have 103 distinct Postal Codes.

In [4]:
len(table_df[table_df.Neighbourhood != 'Not assigned'])

103

Here we can see we have 103 distinct postcodes with neighbourhood(s) assigned.

In [5]:
table_df.shape

(103, 3)

Our dataframe has 103 rows and 3 columns. From above we can deduce that there are no duplicate post codes in the table, nor do any post codes have no assigned neighbourhood(s)

### Adding latitude and the longitude coordinates

My Attempt at geolocating latitude and the longitude coordinates using geopy. This is mostly built on trial and error as some locations could not be found. Latitude and longitude for M7R are from the geospatial csv as geopy couldn't find it using any method.

In [6]:
from geopy.geocoders import Nominatim

geolocator = Nominatim(user_agent="aaa")

from geopy.extra.rate_limiter import RateLimiter
geocode = RateLimiter(geolocator.geocode, min_delay_seconds=1)

table_df['location'] = table_df.Neighbourhood.split(',')[0].split('-')[0].apply(geocode)

AttributeError: 'Series' object has no attribute 'split'

In [ ]:
table_df

In [ ]:
table_df.iloc[76]#.split(',')[0].split('-')[0]

In [ ]:
location = geolocator.geocode('M7R, Toronto, Ontario')
# print(location.latitude, location.longitude)
location

In [ ]:
lat = []

for PC in table_df.Neighbourhood:
    try:
        lat.append(geolocator.geocode('{}, Toronto, Ontario'.format(PC.split(',')[0].split('-')[0])).latitude)
        print(PC)

    except AttributeError:
        lat.append(geolocator.geocode('{}, Toronto, Ontario'.format(PC.split(',')[1].split('-')[0])).latitude)
        print(PC, 'xysssssssssssssssssssssssssssssssssssssssssss')
    except IndexError:
        lat.append(43.6369656)

        
# lat = [geolocator.geocode('{}, Toronto, Ontario'.format(PC.split(',')[0].split('-')[0])).latitude for PC in table_df.Neighbourhood]
# lon

In [ ]:
lat

In [ ]:
lat = ['{}, Toronto, Ontario'.format(PC) for PC in table_df['Postal Code']]
lat = [geolocator.geocode('{}, Toronto, Ontario'.format(PC.split(',')[0].split('-')[0])).latitude for PC in table_df.Neighbourhood]

Using csv provided

In [ ]:
GC = pd.read_csv('Geospatial_Coordinates.csv')
len(GC)